nnlorun.py

```diff
@@ -346,49 +347,74 @@ def copy_to_grid(local_file, grid_file, args, maxrange=MAX_COPY_TRIES):
 
 def grid_copy(infile, outfile, args, maxrange=MAX_COPY_TRIES):
     timeout_secs = 10
-    timeout = datetime.timedelta(minutes=timeout_secs)
+    timeout = datetime.timedelta(seconds=timeout_secs)
+    hash = "MD5"
+    protoc = args.gfaldir.split(":")[0]
+# gfal-sum only returns expected hash for subset of protocols (gsiftp, srm).  Hardcode gsiftp for now.
+    infile_hash_addr = infile.replace(protoc, "gsiftp") 
+    outfile_hash_addr = outfile.replace(protoc, "gsiftp")
+    hashcmd = "{gfal_loc}gfal-sum {{file}} {checksum}".format(gfal_loc=args.gfal_location, checksum=hash)
+    infile_hash = subprocess.check_output(hashcmd.format(file=infile_hash_addr), shell=True, universal_newlines=True).split()[1]
 
     print_flush("Copying {0} to {1}".format(infile, outfile))
-    protoc = args.gfaldir.split(":")[0]
     for protocol in PROTOCOLS: # cycle through available protocols until one works.
         infile_tmp = infile.replace(protoc, protocol)
         outfile_tmp = outfile.replace(protoc, protocol)
+
         print_flush("Attempting Protocol {0}".format(protocol))
         outfile_dir = os.path.dirname(outfile_tmp)
         outfile_fn = os.path.basename(outfile_tmp)
-        lscmd = "{gfal_loc}gfal-ls {dir}".format(dir=outfile_dir, gfal_loc=args.gfal_location)
+        lscmd = "{gfal_loc}gfal-ls {f}".format(f=outfile_hash_addr, gfal_loc=args.gfal_location)
         
         for i in range(maxrange): # try max 10 times for now ;)
-            p = subprocess.check_output(lscmd, shell=True, universal_newlines=True).splitlines()
-            wait_until = datetime.datetime.now() + timeout
-            while outfile_fn in p and datetime.datetime.now() < wait_until:
-                print_flush("Target file {of} already exists.  Removing now.".format(of=outfile_tmp))
-                rmcmd = "gfal-rm {0}".format(outfile_tmp)
-                os.system(rmcmd)
+            try:
                 if debug_level > 1:
-                    print_flush(rmcmd)
-                p = subprocess.check_output(lscmd, shell=True, universal_newlines=True).splitlines()
+                    print_flush(lscmd)
+                p = subprocess.check_output(lscmd, shell=True, universal_newlines=True).splitlines()[0]
+
+                wait_until = datetime.datetime.now() + timeout
+                while outfile_fn in p and datetime.datetime.now() < wait_until:
+                    print_flush("Target file {of} already exists.  Removing now.".format(of=outfile_tmp))
+                    rmcmd = "gfal-rm {0}".format(outfile_tmp)
```

Comment:
Why do we even try to remove the file? Wouldn't a force copy be sufficient?

Reply:
Depends how much you trust gfal...

Reply:
Fair point, however with force copy & a manual checksum test afterwards I trust the result. Have you ever had a protocol where you where able to remove a file but not to force copy it?

I think it should be sufficient if we remove the file after we tried copying with all protocols and never got the right checksum, which definitely shows a copy error. (it would also streamline that bulky copy function)

Reply:
Good stuff; that's what I've done, including an escalating sleep, and duplicated to the other `Xrun.py` files - it works for me in the various failure modes I tried; could you look over and test it?